synapse/visibility.py

```diff
@@ -207,6 +213,15 @@ def allowed(event):
         if erased_senders[event.sender]:
             return prune_event(event)
 
+        retention_policy = retention_policies[event.room_id]
```

Comment:
there are a number of cases where we return the event (either complete, or redacted) before we get here. I think this needs to move earlier?